include/boost/beast/core/async_base.hpp

```diff
@@ -76,54 +76,73 @@ namespace beast {
 
     @code
 
-    // Asynchronously read into a buffer until the buffer is full, or an error occurs
-    template<class AsyncReadStream, class ReadHandler>
-    typename net::async_result<ReadHandler, void(error_code, std::size_t)>::return_type
-    async_read(AsyncReadStream& stream, net::mutable_buffer buffer, ReadHandler&& handler)
+    template<class DeducedHandler, class AsyncReadStream>
+    struct async_read_op
+    : async_base<DeducedHandler, typename AsyncReadStream::executor_type>
     {
-        using handler_type = BOOST_ASIO_HANDLER_TYPE(ReadHandler, void(error_code, std::size_t));
-        using base_type = async_base<handler_type, typename AsyncReadStream::executor_type>;
+        AsyncReadStream& stream_;
+        net::mutable_buffer buffer_;
+        std::size_t total_bytes_transferred_;
+
+        async_read_op(
```

Comment:
Here we need `template<class DeducedHandler>` and `DeducedHandler&&` parameter type

Reply:
why? pass by value-move into place is a perfectly optimisable idiom.